### PR TITLE
Extend FPC Shim support

### PR DESCRIPTION
**Pull request by mbrandenburger**
- add get_signed_proposal
- add get_tx_id
- add get_channel_id
- add get_creator

Signed-off-by: Marcus Brandenburger <bur@zurich.ibm.com>

### CODE REVIEW

**Comment by mbrandenburger** [comment]
TBD what to do with the transient data?

**Comment by mbrandenburger** [comment]
> TBD what to do with the transient data?

transient data support will be separate PR.

**Review by bvavala — CHANGES_REQUESTED** [review]
Two parts: 1) shim extensions and 2) shim internals.

1. Overall, the shim extensions look good. Those can provide the chaincode with useful information which was either previously attested, or extracted from the signed proposal.

Missing: comments and documentation regarding the channel id and, particularly, the creator. The creator is not (and cannot be) validated against the ledger/MSP -- the enclave does not have trustworthy data to do so.

2. The shim internals need some more work. 
First, some data is extracted from the signed proposal, but the signed-proposal signature is not checked. Although documented, this could open to unintended consequences.

Second, the channel id and creator are extracted from the proposal, but the former is already passed to (and attested by) the enclave at init time, while the latter already has a shim API for the name. In both cases, consistency is not checked. Again, this could open to unintended consequences.

File: ecc_enclave/enclave/enclave.cpp [hunk 1]
```diff
@@ -48,6 +51,78 @@ int ecall_cc_invoke(const uint8_t* signed_proposal_proto_bytes,
...
+        // ctx.transient_data = ;
+
+        // set creator
+        common_SignatureHeader signature_header = common_SignatureHeader_init_zero;
+        istream = pb_istream_from_buffer(
+            (const unsigned char*)header.signature_header->bytes, header.signature_header->size);
+        b = pb_decode(&istream, common_SignatureHeader_fields, &signature_header);
+        COND2LOGERR(!b, PB_GET_ERROR(&istream));
+
+        ctx.creator = ByteArray(signature_header.creator->bytes,
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply ecc_enclave/enclave/enclave.cpp:115]: right - I figured out how to extract creator_name directly from the signed proposal, so I jusggest just to remove the ocall. WIP

File: ecc_enclave/enclave/enclave.cpp [hunk 2]
```diff
@@ -48,6 +51,78 @@ int ecall_cc_invoke(const uint8_t* signed_proposal_proto_bytes,
...
+
+        // set tx_id and channel id
+        common_ChannelHeader channel_header = common_ChannelHeader_init_zero;
+        istream = pb_istream_from_buffer(
+            (const unsigned char*)header.channel_header->bytes, header.channel_header->size);
+        b = pb_decode(&istream, common_ChannelHeader_fields, &channel_header);
+        COND2LOGERR(!b, PB_GET_ERROR(&istream));
+
+        ctx.tx_id = std::string(channel_header.tx_id);
+        ctx.channel_id = std::string(channel_header.channel_id);
```
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply ecc_enclave/enclave/enclave.cpp:89]: right! added

**Comment by mbrandenburger** [comment]
I just experienced that this code is not compatible with Idemix.

With current FPC, Idemix is working as long as you not call `get_creator_name`. The problem with Idemix is that the serialized identity does not contain a x509 cert (in pem format), instead it contains a serialized idemix identity (as protobuf msg). As said, this is only a problem with `get_creator_name`, where we try to extract the cert subject.

That is, the code in this PR fails when it tries to parse the x509 as it is actually an idemix identity.

**Comment by mbrandenburger** [comment]
> I just experienced that this code is not compatible with Idemix.
> 
> With current FPC, Idemix is working as long as you not call `get_creator_name`. The problem with Idemix is that the serialized identity does not contain a x509 cert (in pem format), instead it contains a serialized idemix identity (as protobuf msg). As said, this is only a problem with `get_creator_name`, where we try to extract the cert subject.
> 
> That is, the code in this PR fails when it tries to parse the x509 as it is actually an idemix identity.

**Comment by mbrandenburger** [comment]
@bvavala are you still looking at this?

**Comment by bvavala** [comment]
comments addressed - not tested in hw mode

**Pull request merged**